packages/plugin-ext/src/main/browser/view/tree-view-widget.tsx

```diff
@@ -546,6 +559,143 @@ export class TreeViewWidget extends TreeViewWelcomeWidget {
+        if (event.dataTransfer) {
+            const canDrop = event.dataTransfer.types.some(type => event.dataTransfer.types.indexOf(type) >= 0) ||
+                event.dataTransfer.types.indexOf(this.treeDragType) > 0 ||
+                this.options.dropMimeTypes!.indexOf('files') > 0 && event.dataTransfer.files.length > 0;
+            if (canDrop) {
+                event.preventDefault();
+                event.dataTransfer.dropEffect = 'move';
+            } else {
+                event.dataTransfer.dropEffect = 'none';
+            }
+            event.stopPropagation();
+        }
+    }
+
+    protected handleDropEvent(node: TreeViewNode | undefined, event: React.DragEvent<HTMLElement>): void {
+        if (event.dataTransfer) {
+            const items: [string, string | DataTransferFileDTO][] = [];
+            let files: string[] = [];
+            try {
+                for (let i = 0; i < event.dataTransfer.items.length; i++) {
+                    const transferItem = event.dataTransfer.items[i];
+                    if (transferItem.type !== this.treeDragType) {
+                        // do not pass the artificial drag data to the extension
+                        const f = event.dataTransfer.items[i].getAsFile();
+                        if (f) {
+                            const fileId = this.dndFileContentStore.addFile(f);
+                            files.push(fileId);
+                            const uri = f.path ? {
+                                scheme: 'file',
+                                path: f.path,
+                                authority: '',
+                                query: '',
+                                fragment: ''
+                            } : undefined;
+                            items.push([transferItem.type, new DataTransferFileDTO(f.name, fileId, uri)]);
+                        } else {
+                            const textData = event.dataTransfer.getData(transferItem.type);
+                            if (textData) {
+                                items.push([transferItem.type, textData]);
+                            }
+                        }
+                    }
+                }
+                if (items.length > 0 || event.dataTransfer.types.indexOf(this.treeDragType) >= 0) {
+                    event.preventDefault();
+                    event.stopPropagation();
+                    const filesCopy = [...files];
+                    this.model.proxy?.$drop(this.id, node?.id, items, CancellationToken.None).then(() => {
+                        for (const file of filesCopy) {
+                            this.dndFileContentStore.removeFile(file);
+                        }
+                    });
+                    files = [];
+                }
+            } finally {
+                for (const file of files) {
+                    this.dndFileContentStore.removeFile(file);
+                }
+            }
```

Comment:
I'm a little confused about the logic here. It looks like _if_ we've gotten items, or we ran into an error before line 680, then we want to wait before removing the files from the store; otherwise, we want to remove them immediately?

If the call to `.$drop` throws an error, we will never clear the files, because we will have run 689, but we'll never get to 685-688.

If I'm right about the goal, then we could achieve the same end by `awaiting` at line 684 and not bothering with `filesCopy`. Then failure of the `.$drop()` call would be covered by the `try` and we'd always end up in the `finally`, but only after we'd either succeeded or failed in the `.$drop()`, or skipped it because there was nothing to do.

Reply:
I'm reluctant to make this an async method, as it's an event handler. It's just to easy to confuse what's happening now and what's later. But you're right about the logic being broken. I'll try to find a more appropriate form.